File: Coding_231215 1D_DP/최원준/최원준/2번.py
# https://leetcode.com/problems/house-robber-ii/description/

'''
1. 아이디어 :
    dp로 풀 수 있다
2. 시간복잡도 :
    O(n)
3. 자료구조 :
    배열
'''


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        nums1 = nums[1:]
        nums2 = nums[:-1]

        def dynamic(nums):
            prev, curr = 0, 0
            for n in nums:
                cmax = max(n + prev, curr)
                prev = curr
                curr = cmax
            return cmax

        return max(dynamic(nums1), dynamic(nums2))

File: Coding_231215 1D_DP/최원준/이강훈/2번.java
/*
https://leetcode.com/problems/house-robber-ii/description/

아이디어 ( 솔루션 보고 품 )
1시작 , 2시작 dp 배열 2개 생성
 -> 1시작은 마지막 집을 사용 못하므로 length - 2 , 2시작은 마지막 집 사용 가능하므로 length - 1 .. 2개 중 큰 값이 정답

자료구조
배열

시간복잡도
O (N)
*/

import java.util.Arrays;

public class Solution {
    public int rob(int[] nums) {
        if (nums.length <= 2) return Arrays.stream(nums).max().getAsInt();

        int[] dyFromOne = new int[nums.length];
        int[] dyFromTwo = new int[nums.length];

        dyFromOne[0] = nums[0];
        dyFromOne[1] = nums[0];

        dyFromTwo[0] = 0;
        dyFromTwo[1] = nums[1];

        for (int i = 2; i < nums.length; i++) {
            dyFromOne[i] = Math.max(dyFromOne[i-1], dyFromOne[i-2] + nums[i]);
            dyFromTwo[i] = Math.max(dyFromTwo[i-1], dyFromTwo[i-2] + nums[i]);
        }

        return Math.max(dyFromOne[nums.length - 2], dyFromTwo[nums.length - 1]);
    }
}
